frame/elections/src/lib.rs

```diff
@@ -890,11 +913,72 @@ impl<T: Config> Pallet<T> {
 		debug_assert!(_remainder.is_zero());
 	}
 
-	/// Run the phragmen election with all required side processes and state updates, if election
-	/// succeeds. Else, it will emit an `ElectionError` event.
+	/// Run an election with all required side processes and state updates, if election
+	/// succeeds. Else, it will emit an `ElectionError` event. The election algorithm is defined
+	/// by the implementor of `Self::ElectionSolver`.
 	///
 	/// Calls the appropriate [`ChangeMembers`] function variant internally.
-	fn do_phragmen() -> Weight {
+	fn do_election() -> Weight {
+		let PreElectionResults {
+			num_to_elect,
+			candidate_ids,
+			candidates_and_deposit,
+			voters_and_stakes,
+			voters_and_votes,
+			num_edges,
+		} = match Self::do_pre_solve_election() {
+			Ok(results) => results,
+			Err(event) => match event {
+				Event::EmptyTerm => {
+					Self::deposit_event(event);
+					return T::DbWeight::get().reads(3)
+				},
+				Event::ElectionError => {
+					Self::deposit_event(event);
+					log::error!(
+						target: "runtime:elections",
+						"Failed to run election. Number of voters exceeded",
+					);
+					let max_voters = <T as Config>::MaxVoters::get() as usize;
+					return T::DbWeight::get().reads(3 + max_voters as u64)
+				},
+				_ => unreachable!("should not happen"),
+			},
+		};
+
+		let num_candidates = candidates_and_deposit.len() as u32;
+		let num_voters = voters_and_votes.len() as u32;
+		let num_edges = num_edges;
+
+		let election_result =
+			T::ElectionSolver::solve(num_to_elect, candidate_ids, voters_and_votes)
+				.map(
+					|ElectionResult::<
+						T::AccountId,
+						<T::ElectionSolver as NposSolver>::Accuracy,
+					> {
+					     winners,
+					     assignments: _,
+					 }| winners,
+				)
+				.map_err(|e| {
+					log!(warn, "Failed to run election [{:?}].", e);
+					Self::deposit_event(Event::ElectionError);
+				});
+
+		let pre_and_election_weight =
+			T::ElectionSolver::weight::<T::SolverWeightInfo>(num_voters, num_candidates, num_edges)
+				.saturating_add(Weight::zero()); // replace with weights::pre_solve_election_weight();
+
+		match election_result {
+			Ok(winners) =>
+				Self::do_post_solve_election(winners, candidates_and_deposit, voters_and_stakes)
+					.saturating_add(pre_and_election_weight),
+			Err(_) => pre_and_election_weight,
+		}
+	}
+
+	fn do_pre_solve_election() -> Result<PreElectionResults<T>, Event<T>> {
```

Comment:
Returning an event is pretty strange. You should probably stick to returning an instance of `Error<T>` here and treat it in the upper function.

Reply:
I agree. Changing to return `Error`s instead